pulsar-transaction/coordinator/src/main/java/org/apache/pulsar/transaction/coordinator/impl/TxnLogBufferedWriter.java

```diff
@@ -0,0 +1,483 @@
+        prefix.writeShort(BATCHED_ENTRY_DATA_PREFIX_VERSION);
+        ByteBuf actualContent = this.dataSerializer.serialize(this.dataArray);
+        ByteBuf pairByteBuf = Unpooled.wrappedUnmodifiableBuffer(prefix, actualContent);
+        // Flush.
+        managedLedger.asyncAddEntry(pairByteBuf, this, flushContext);
+        // Clear buffers.
+        this.dataArray.clear();
+        this.flushContext = FlushContext.newInstance();
+        this.bytesSize = 0;
+    }
+
+    /**
+     * see {@link AsyncCallbacks.AddEntryCallback#addComplete(Position, ByteBuf, Object)}.
+     */
+    @Override
+    public void addComplete(Position position, ByteBuf entryData, Object ctx) {
+        final FlushContext flushContext = (FlushContext) ctx;
+        try {
+            final int batchSize = flushContext.asyncAddArgsList.size();
+            for (int batchIndex = 0; batchIndex < batchSize; batchIndex++) {
+                final AsyncAddArgs asyncAddArgs = flushContext.asyncAddArgsList.get(batchIndex);
+                BitSetRecyclable bitSetRecyclable = BitSetRecyclable.create();
+                bitSetRecyclable.set(batchIndex);
+                long[] ackSet = bitSetRecyclable.toLongArray();
+                bitSetRecyclable.recycle();
+                final TxnBatchedPositionImpl txnBatchedPosition = new TxnBatchedPositionImpl(position, batchSize,
+                        batchIndex, ackSet);
+                // Because this task already running at ordered task, so just "run".
+                SafeRunnable.safeRun(() -> {
+                    asyncAddArgs.callback.addComplete(txnBatchedPosition, asyncAddArgs.ctx);
+                }).run();
+            }
+        } finally {
+            entryData.release();
+            flushContext.recycle();
+        }
+    }
+
+    /**
+     * see {@link AsyncCallbacks.AddEntryCallback#addFailed(ManagedLedgerException, Object)}.
+     */
+    @Override
+    public void addFailed(ManagedLedgerException exception, Object ctx) {
+        final FlushContext flushContext = (FlushContext) ctx;
+        try {
+            for (int i = 0; i < flushContext.asyncAddArgsList.size(); i++) {
+                final AsyncAddArgs asyncAddArgs = flushContext.asyncAddArgsList.get(i);
+                // Because this task already running at ordered task, so just "run".
+                SafeRunnable.safeRun(() -> asyncAddArgs.callback.addFailed(exception, asyncAddArgs.ctx)).run();
+            }
+        } finally {
+            flushContext.recycle();
+        }
+    }
+
+    /**
+     * Release resources and executing tasks.
+     */
+    @Override
+    public void close() {
+        this.state = State.CLOSING;
+        singleThreadExecutorForWrite.execute(() -> {
+            if (!flushContext.asyncAddArgsList.isEmpty()) {
+                doTrigFlush(true);
+            }
+        });
+        if (this.scheduledFuture != null) {
+            this.scheduledFuture.cancel(false);
```

Comment:
We should check if the `scheduledFuture` is able to cancel first? If we got an exception here, the state will stay at CLOSING state.